python/hail/expr/functions.py

```diff
@@ -43,8 +47,9 @@ def null(t):
     """
     return construct_expr(Literal('NA: {}'.format(t._jtype.parsableString())), t)
 
+
 @typecheck(x=anytype, dtype=nullable(hail_type))
-def literal(x, dtype=None):
+def literal(x: Any, dtype: Optional[Union[HailType, str]] = None) -> Expression:
```

Comment:
I think I prefer the equivalent `Union[HailType, str, None]`, but not strongly. Up to you.

Reply:
I prefer Optional, because it seems to better represent the branch logic -- either it's defined or it's not, and if it's defined, then either it's a str or a Type. I don't feel strongly, though.

Reply:
Ah, another thing -- this would also let us abstract it out later more easily by defining HT as `Union[HailType, str]` once.

Reply:
Yeah, let's go with yours.

Reply:
Good point.